Library/Homebrew/version.rb

```diff
@@ -394,54 +393,27 @@ def self._parse(spec)
 end
 
 class HeadVersion < Version
-  attr_reader :version
   attr_reader :commit
 
   def initialize(val)
     if val.respond_to?(:to_str)
-      @version = if (val.to_str.start_with?("HEAD"))
-        val.to_str
-      else
-        "HEAD-#{val}"
-      end
-      @commit = @version[/HEAD-(.+)/, 1]
+      @version = val.to_str
     else
       raise TypeError, "Version value must be a string; got a #{val.class} (#{val})"
     end
+    @commit = @version[/HEAD-(.+)/, 1]
   end
 
   def update_commit(commit)
     @commit = commit
-    update_version
-    version
-  end
-
-  def to_s
-    if commit
+    @version = if commit
       "HEAD-#{commit}"
     else
       "HEAD"
     end
   end
 
-  def <=>(other)
-    return unless other.is_a? Version
-    if other.is_a? HeadVersion
-      version <=> other.version
-    end
-  end
-
   def head?
     true
   end
-
-  private
-
-  def update_version
-    @version = if commit
-      "HEAD-#{commit}"
-    else
-      "HEAD"
-    end
-  end
```

Comment:
@vladshablinsky These kind of discarded changes should be removed from git history by squashing the commit.